subtests/docker_cli/systemd/systemd.py

```diff
@@ -5,82 +5,124 @@
         self.sysd_action('daemon-reload')
-        self.sysd_action('start', self.config['sysd_unit_file'])
-        utils.wait_for(lambda: not is_port_free(port, host), 10)
-        time_from_socket = self.read_socket(host, port)
-        self.failif(not time_from_socket.isdigit(),
-                    "Data received from container is non-numeric: '%s'"
-                    % time_from_socket)
+        for act in 'start', 'status':
+            self.sysd_action(act, self.config['sysd_unit_file'])
 
     def cleanup(self):
-        super(systemd, self).cleanup()
-        filepath = self.config['sysd_unitf_dest']
+        super(systemd_base, self).cleanup()
+        filepath = self.sub_stuff['unitf_dst']
         if exists(filepath):
             self.sysd_action('stop', self.config['sysd_unit_file'])
             unlink(filepath)
             self.sysd_action('daemon-reload')
-        DockerImages(self).clean_all([self.config['name']])
+        DockerImages(self).clean_all([self.config['image_name']])
 
     @staticmethod
-    def sysd_action(action=None, sysd_unit=None):
+    def sysd_action(action, sysd_unit=None):
         command = 'systemctl {}'.format(action)
         if sysd_unit:
             command = '{} {}'.format(command, sysd_unit)
         utils.run(command, ignore_status=False)
 
-    @staticmethod
-    def read_socket(host, port):
-        max_read_tries = 10
-        num_read_try = max_read_tries
-        while num_read_try:
-            # using timeout 13-sec as it's 1-sec longer than two DNS timeouts
-            s = socket.create_connection((host, port), 13)
-            s.settimeout(2.0)
-            try:
-                time_from_socket = s.recv(512).strip()
-                if time_from_socket:
-                    s.shutdown(socket.SHUT_RDWR)
-                    return time_from_socket
-                num_read_try -= 1
-            except socket.timeout:
-                pass
-            time.sleep(1.0)
-        raise DockerTestError("Failed to read from socket with %d tries"
-                              % max_read_tries)
+    def write_unitfile(self, vars_dict):
+        tmpdirf = self.sub_stuff['tmpdirf']
+        tmp_write = '{}.tmp'.format(tmpdirf)
+        with open(tmpdirf, 'r') as srcfile:
+            lines = srcfile.readlines()
+            # if file is a unitfile, prepend line with 'Environment='
+            if '.service' in tmpdirf:
+                prep = 'Environment='
+            else:
+                prep = ''
+            # edit lines with matching string to dict keys
+            for key in vars_dict.keys():
+                ndl_found = next((ndl for ndl in lines if key in ndl), None)
+                if ndl_found is not None:
+                    indx = lines.index(ndl_found)
+                    lines[indx] = '{}{}{}\n'.format(prep, key, vars_dict[key])
+        with open(tmp_write, 'w') as outfile:
+            outfile.writelines(lines)
+        copyfile(tmp_write, tmpdirf)
```

Comment:
Full agreement with @cevich : this has grown waaaaaaaay too complex for what it does. And in addition:

  * The moment you started using this for rewriting Dockerfile, you should've renamed the function: Dockerfile isn't a unitfile, so the function name is now highly misleading.
  * The `copyfile` at the end is super dangerous: it leaves the tmp file behind. You don't want that. You don't want to leave any trace. That should be `os.rename()`.
  * the 'Environment=' was fine when this was write_unitfile but now it's dangerous: since this is no longer a function dedicated to unit files, this function should not know about Environment=. That should be left to the caller now.

One possible way to do this (completely untested, it's late and I'm tired):

    def sed_file_in_place(path, replacements):
        tmpfile = path + '.tmp'
        os.unlink(tmpfile)           # in case there's a dropping from a previous run
        s = open(path).read()
        for key, newvalue in replacements:
            searchfor = re.compile(key + '.*$', re.MULTILINE)
            s = re.sub(searchfor, key + newvalue, s)
        ...writing to tmpfile, then os.rename(), left as exercise for the reader